hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/mongodb/MongoSqlConnectorBase.java

```diff
@@ -197,57 +170,65 @@ protected static ProcessorMetaSupplier wrap(DagBuildContext ctx, ProcessorSuppli
-                                          .map(e -> e.unwrap(RexNode.class).accept(visitor))
-                                          .filter(proj -> proj instanceof String)
-                                          .map(p -> (String) p)
-                                          .collect(toList());
-
-            if (fields.isEmpty()) {
-                throw new IllegalArgumentException("Projection list cannot be empty");
-            }
-            if (fields.size() != projectionNodes.size()) {
-                return new TranslationResult<>(allFieldsExternalNames(context.getTable()), false);
+    private static List<ProjectionData> translateProjections(
+            List<HazelcastRexNode> projectionNodes,
+            DagBuildContext context,
+            RexToMongoVisitor visitor
+    ) {
+        List<ProjectionData> projection = new ArrayList<>();
+
+        MongoTable table = context.getTable();
+        String[] externalNames = table.externalNames();
+        for (int i = 0; i < projectionNodes.size(); i++) {
+            Object translated = projectionNodes.get(i).unwrap(RexNode.class).accept(visitor);
+            InputRef ref = InputRef.match(translated);
+            if (ref != null) {
+                String externalName = externalNames[ref.getInputIndex()];
+                Document projectionExpr = bsonToDocument(Projections.include(externalNames));
+                projection.add(new ProjectionData(externalName, projectionExpr, i, table.fieldType(externalName)));
+            } else {
+                Document projectionExpr = new Document("projected_value_" + i, new Document("$literal", translated));
+                projection.add(new ProjectionData("projected_value_" + i, projectionExpr, i, null));
             }
+        }
 
-            return new TranslationResult<>(fields, true);
-        } catch (Throwable t) {
-            return new TranslationResult<>(allFieldsExternalNames(context.getTable()), false);
+        if (projection.isEmpty()) {
+            throw new IllegalArgumentException("Projection list cannot be empty");
         }
-    }
 
-    private static List<String> allFieldsExternalNames(MongoTable table) {
-        return table.getFields().stream()
-                    .map(f -> ((MongoTableField) f).externalName)
-                    .collect(toList());
+        return projection;
     }
 
-    static final class TranslationResult<T> {
-        final T result;
-        final boolean allProceeded;
-
-        private TranslationResult(T result, boolean allProceeded) {
-            this.result = result;
-            this.allProceeded = allProceeded;
+    @Override
+    public boolean supportsExpression(@Nonnull HazelcastRexNode expression) {
+        RexToMongoVisitor visitor = new RexToMongoVisitor();
+        RexNode rexNode = expression.unwrap(RexNode.class);
+        try {
+            rexNode.accept(visitor);
+            return true;
+        } catch (Throwable ignored) {
+            return false;
```

Comment:
@TomaszGaweda  exception-driven control flow?

Reply:
Yes - if something is unsupported, visitor will throw an UnsupportedOperationException, which means we cannot parse this part